### PR TITLE
fix crash in libusb callback handling

### LINKED ISSUE #2429: Fix crash in libusb callback handling
### Bug Report
Fix crash in libusb callback handling

Preconditions:
SDLCore and HMI are started.
Start SDL with command ./smartDeviceLinkCore | grep -e '.eart.eat' to check heartbeat only

##### Reproduction Steps
Start application with "HeartBeat" is switched on
Wait 30 minutes 

##### Expected Behavior
1.Application is registered
2. Heartbeat is sent
3. Connection is not closed by HB timeout reason.
4. No core dumps 

##### Observed Behavior
Connection is closed with core dump

##### OS & Version Information
* OS/Version: 
* SDL Core Version: 
* Testing Against:

### PR DESCRIPTION
**Pull request by Ypostolov**
Fixes #2429

This PR is ready for review.

### Risk
This PR makes no API changes.

### Testing Plan
Changes in PR are already covered within existing unit tests.

### Summary
The rootcause of the issue is the callback function OnOutTransfer().
This is occuring when when AbortConnection() is called or
UsbConnection object is destroyed. In case if have OutTransfer
currently taking place we cancel this transfer with
libusb_cancel_transfer(). So in order to fix the issue we capture this status in our callback
function, free allocated transfer structure used by libusb and exit not
touching anything else. It's also taken under consideration

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

File: src/components/transport_manager/src/usb/libusb/usb_connection.cc [hunk 1]
```diff
@@ -206,6 +206,14 @@ bool UsbConnection::PostOutTransfer() {
 
 void UsbConnection::OnOutTransfer(libusb_transfer* transfer) {
   LOG4CXX_TRACE(logger_, "enter with  Libusb_transfer*: " << transfer);
+
+  if (transfer->status == LIBUSB_TRANSFER_CANCELLED) {
+    LOG4CXX_DEBUG(logger_,"Free already canceled transfer.");
+    libusb_free_transfer(transfer);
+    LOG4CXX_TRACE(logger_,"exit");
```
Comment (AKalinich-Luxoft) [inline src/components/transport_manager/src/usb/libusb/usb_connection.cc]: @Ypostolov please add `LOG4CXX_AUTO_TRACE` at line 208 and remove this one at 213

Comment (Ypostolov) [reply src/components/transport_manager/src/usb/libusb/usb_connection.cc]: fixed in 8977c019208ef0b106eb7361f53f64d825ad4682

File: src/components/transport_manager/src/usb/libusb/usb_connection.cc [hunk 2]
```diff
@@ -205,7 +205,15 @@ bool UsbConnection::PostOutTransfer() {
 }
 
 void UsbConnection::OnOutTransfer(libusb_transfer* transfer) {
-  LOG4CXX_TRACE(logger_, "enter with  Libusb_transfer*: " << transfer);
+  LOG4CXX_AUTO_TRACE(logger_);
+  LOG4CXX_DEBUG(logger_,"enter with Libusb_transfer*: " << transfer);
```
Comment (LuxoftAKutsan) [inline src/components/transport_manager/src/usb/libusb/usb_connection.cc]: @Ypostolov this is at least trace. But I suppose that transfer pointer is redundant info in logs

Comment (Ypostolov) [reply src/components/transport_manager/src/usb/libusb/usb_connection.cc]: Fixed in 3c539a8a53cfd07be342e9a557446b71d8247c5a

**Review by iCollin — COMMENTED** [review]
@Ypostolov could you please rebase on latest develop. Thank you!

File: src/components/transport_manager/src/usb/libusb/usb_connection.cc.orig [hunk 3]
```diff
@@ -0,0 +1,420 @@
+/*
```
Comment (iCollin) [inline src/components/transport_manager/src/usb/libusb/usb_connection.cc.orig:1]: this file can be removed, correct?

File: src/components/transport_manager/src/usb/libusb/usb_connection.cc [hunk 4]
```diff
@@ -200,6 +207,13 @@ TransportAdapter::Error UsbConnection::PostOutTransfer() {
 }
 
 void UsbConnection::OnOutTransfer(libusb_transfer* transfer) {
+  SDL_LOG_TRACE("enter");
```
Comment (iCollin) [inline src/components/transport_manager/src/usb/libusb/usb_connection.cc:210]:
```suggestion
  SDL_LOG_AUTO_TRACE();
```

File: src/components/transport_manager/src/usb/libusb/usb_connection.cc [hunk 5]
```diff
@@ -128,6 +128,13 @@ std::string hex_data(const unsigned char* const buffer,
 }
 
 void UsbConnection::OnInTransfer(libusb_transfer* transfer) {
+  SDL_LOG_TRACE("enter");
```
Comment (iCollin) [inline src/components/transport_manager/src/usb/libusb/usb_connection.cc:131]:
```suggestion
  SDL_LOG_AUTO_TRACE();
```

File: src/components/transport_manager/src/usb/libusb/usb_connection.cc.orig (hunk 3 above)
Comment (Ypostolov) [reply src/components/transport_manager/src/usb/libusb/usb_connection.cc.orig]: @iCollin Fixed in 4f9ac4004124a2b723a8c307b3b2023779d17e38. Please take a look.

**Review by iCollin — APPROVED** [review]

**Pull request merged**